File: core/src/main/java/fr/gael/dhus/service/EvictionService.java
/*
 * Data Hub Service (DHuS) - For Space data distribution.
 * Copyright (C) 2013,2014,2015 GAEL Systems
 *
 * This file is part of DHuS software sources.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package fr.gael.dhus.service;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import org.joda.time.DateTime;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Propagation;
import org.springframework.transaction.annotation.Transactional;

import fr.gael.dhus.database.dao.ProductDao;
import fr.gael.dhus.database.dao.interfaces.IEvictionDao;
import fr.gael.dhus.database.object.Eviction;
import fr.gael.dhus.database.object.Product;
import fr.gael.dhus.datastore.eviction.EvictionManager;
import fr.gael.dhus.datastore.eviction.EvictionStrategy;

@Service
public class EvictionService extends WebService
{
   @Autowired
   private EvictionManager evictionMgr;
   
   @Autowired
   private IEvictionDao evictionDao;
   
   @Autowired
   private ProductDao productDao;
   
   @PreAuthorize ("hasRole('ROLE_SYSTEM_MANAGER')")
   @Transactional (readOnly=true, propagation=Propagation.REQUIRED)
   public int getKeepPeriod()
   {
      return evictionDao.getEviction ().getKeepPeriod ();
   }

   @PreAuthorize ("hasRole('ROLE_SYSTEM_MANAGER')")
   @Transactional (readOnly=true, propagation=Propagation.REQUIRED)
   public int getMaxDiskUsage()
   {
      return evictionDao.getEviction ().getMaxDiskUsage ();
   }

   @PreAuthorize ("hasRole('ROLE_SYSTEM_MANAGER')")
   @Transactional (readOnly=true, propagation=Propagation.REQUIRED)
   public EvictionStrategy getStrategy()
   {
      return evictionDao.getEviction ().getStrategy ();
   }

   @PreAuthorize ("hasRole('ROLE_SYSTEM_MANAGER')")
   @Transactional (readOnly=false, propagation=Propagation.REQUIRED)
   public void save (EvictionStrategy strategy, int keep_period,
         int max_disk_usage)
   {
      evictionDao.update (strategy, keep_period, max_disk_usage);
      evictionMgr.computeNextProducts ();
   }

   @PreAuthorize ("hasRole('ROLE_SYSTEM_MANAGER')")
   @Transactional (readOnly=true, propagation=Propagation.REQUIRED)
   public List<Product> getEvictableProducts()
   {
      return new ArrayList<> (evictionMgr.getProducts ());
   }
   
   @PreAuthorize ("hasRole('ROLE_SYSTEM_MANAGER')")
   @Transactional (readOnly=false, propagation=Propagation.REQUIRED)
   public void doEvict()
   {
      evictionMgr.doEvict ();
   }
   
   @PreAuthorize ("isAuthenticated ()")
   @Transactional (readOnly=true, propagation=Propagation.REQUIRED)
   @Cacheable (value = "product_eviction_date", key = "#pid")
   public Date getEvictionDate (Long pid)
   {
      Eviction eviction = evictionDao.getEviction (); 
      if (eviction.getStrategy () == EvictionStrategy.NONE)
      {
         return null;
      }
      Product p = productDao.read (pid);
      DateTime dt = new DateTime (p.getIngestionDate ());
      DateTime res = dt.plusDays (eviction.getKeepPeriod ());
      return res.toDate ();
   }

   // Methods for unit tests
   void setEvictionDao (IEvictionDao eviction_dao)
   {
      this.evictionDao = eviction_dao;
   }

   void setEvictionMgr (EvictionManager eviction_mgr)
   {
      this.evictionMgr = eviction_mgr;
   }
}

File: client/webclient/src/main/frontend/app/scripts/services/admin/eviction-service.js
/* 
 * Data HUb Service (DHuS) - For Space data distribution.
 * Copyright (C) 2013,2014,2015,2016 European Space Agency (ESA)
 * Copyright (C) 2013,2014,2015,2016 GAEL Systems
 * Copyright (C) 2013,2014,2015,2016 Serco Spa
 * 
 * This file is part of DHuS software sources.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
'use strict';

angular
  .module('DHuS-webclient')
.factory('EvictionService', function($http, $q){
  var service = {
    model: null,
    evictionUrl: "api/stub/admin/evictions",
    evictionProductsUrl: "api/stub/admin/evictions/products",
    getEviction: function(){
      var self = this;
      if(self.model)
      {
          var defer = $q.defer();
          defer.resolve(self.model);
          return defer.promise;

      }else{
        return $http(
        {
          url: ApplicationConfig.baseUrl + self.evictionUrl, 
          method: "GET"
        })
          .then(function(response) {
            return (response.status == 200)?response.data:[];
          });
      }
    },

    updateEviction: function(model){
      var self = this;
      return $http(
        {
          url: ApplicationConfig.baseUrl + self.evictionUrl, 
          method: "PUT",
          data:model
        })
          .success(function(response) {
            self.model = angular.copy(model);            
          })
          .error(function(){
          });
    },

    runEviction: function(model){
      var self = this;
      return $http(
        {
          url: ApplicationConfig.baseUrl + self.evictionUrl + "?run=true", 
          method: "PUT",
          data:model
        });
    },

    getEvictionProducts: function(){
      var self = this;
      return $http(
        {
          url: ApplicationConfig.baseUrl + self.evictionProductsUrl, 
          method: "GET"
        })
          .then(function(response) {
            return (response.status == 200)?response.data:[];
          }); 
    },
    init: function(){
      var self = this;
      this.getEviction()
        .then(function(model){
          self.model = model;
        });
    }
  };
  service.init();

  return service;
   
});



